Expose p2p connection of gateways and server (#150)

nondetalle opened the pull request:
This pull request is meant to fix the problem brought up in issue #150

`NetworkServerHelper::Install` used to implicitly create a point-to-point connection between gateways and the network server. Specifically, this function would install a `PointToPointNetDevice` on gateways, then required by the `ForwarderHelper` application installer on gateways. Thus, if `ForwarderHelper::Install` was called before `NetworkServerHelper::Install` in a custom simulation, it would generate a zero pointer runtime error with a non-descriptive message. This can be especially difficult to debug for a new user (happened to me as well in the past).

Given that NetDevice installation is usually meant to happen in the main ns-3 simulation file, I propose to move it there. The idea is to make it explicit such that new users avoid committing the error described above. I also added clearer error messages via assertions requiring a `PointToPointNetDevice` to be present when calling `ForwarderHelper::Install`. Currently, the module only supports a p2p connection, later on we could extend this to support any type of connection between gateways and the network server.

nondetalle committed:
Expose p2p connection of gateways to server (#150)

pagmatt commented on helper/network-server-helper.cc:
```diff
@@ -86,28 +82,34 @@ Ptr<Application>
+                      "Connection with gateways is not PointToPoint");
+        for (uint32_t j = 0; j < channel->GetNDevices(); ++j)
+        {
+            Ptr<Node> gwNode = channel->GetDevice(j)->GetNode();
```
If I understand correctly, you are looping through the channel endpoint devices (which should be 2 given that it is a P2P channel?). Therefore, this `gwNode` is not necessarily the gateway, but it could be either the actual gateway, or the network server node. If this is correct, please rename the corresponding variables to reflect this.                                                            Anyway, I am not a huge fan of this solution since 1) it assumes that the link is indeed P2P, which seems to be an assumption which will not hold with future developments and 2) it is quite convoluted. What about adding a `NetworkServerHelper` method that accepts as parameters the gateway nodes and the corresponding P2P devices, and stores these pairs to then pass to `AddGateway` ? For instance, something like` RegisterGatewayP2P(Node gw, NetDevice p2p)` to be called when installing the P2P link

nondetalle replied:
Ideally, the server should not have to retrieve the gateway's node. This happens because the packet scheduling features that in real networks happen in gateways, here are implemented in the server. So, currently the server - with a notable abstraction jump - directly gets the gateway's MAC layer from the node to access information (mainly, availability for next transmission) used to schedule a packet. In reality, the server should send packets to gateways with an attached timestamp and gateways should take care of scheduling the radio transmission just-in-time.

My line of though was to try and automate the GW node's retrieval, hiding stuff that may be changed in the future. That said, I am conflicted on how to proceed. On one hand, I like your solution because it makes explicit from a user perspective that we currently only support P2P. On the other, I do not like passing the NetDevice, as it is an object already accessible through the NS node... Mind that we are talking of **the P2P net device of the NS** connected to the GW we are passing the node of, not the one of the GW. It seems to me that with `RegisterGatewayP2P(Node gw, NetDevice p2p)` we would just move the convoluted part in the main.cc file.

For clarity: it looked cleaner beforehand because the P2P connection was instantiated by the NS helper, thus having GW's node and NS's P2P net device in the same place. This was causing the bug described on top of this thread, because users did not know where the P2P instantiation was happening. It gets convoluted because we are trying to retrieve something we are not supposed to (the GW's node, and the MAC later on) from the NS.

nondetalle committed:
Expose gw registration to simplify NetworkServerHelper::Install
+ remove NodeContainer install of NetworkServerHelper
+ add P2PGwRegistration_t helper structure to facilitate independent instantiation of P2P and server app

nondetalle commented:
Hello @pagmatt,

Here is a new version, I tried to make it as clean as I could get it on my own. I rebased it onto develop, but you can still  easily check the latest changes because they are all in commit 9ab8421. I followed your suggestion to try and have a sort of `RegisterGatewayP2P` function to be used in the main simulation files. As I mentioned in my last reply, this moves a bit of complexity in the main files (see the examples) but I do not think it is too bad.

In summary, the changes proposed in this PR are meant to accomplish the following: 
- Expose the P2P connection between GWs and server to prevent the bug that happened if the gateways' apps were installed before the server's app (highlighted in issue #150). I feel it is also correct in theory because you should not be forced to have a specific order of installation of independent applications on different nodes;
- Allow users to instantiate the P2P connection before installing applications, which is more in line with ns-3's way of doing things (i.e., that you should normally instantiate things in the order of nodes -> net devices -> applications, being completely separate steps);
- Clarify that the module currently only supports P2P connections internally by forcing you to use `AddGatewayP2P` to register gateways connected to the network server.

Other  minor changes coming with the last commit are:
- removing the option to provide a NodeContainer to NetworkServerHelper. Installing the app on multiple nodes does not make a lot of sense given that you have to provide net devices created beforehand.  In most cases one server is enough but people who want to play with more of them can always create different NetworkServerHelpers and register different devices/gateways to them;
- adding a data structure to help storing what's needed for gateway registration in the NS app. It makes easier to have the P2P connection in one part of your code without depending on whether the NetworkServerHelper has been created yet.

Let me know your thoughts.

pagmatt commented on helper/network-server-helper.h:
```diff
@@ -81,12 +91,11 @@ class NetworkServerHelper
     ObjectFactory m_factory;
 
-    NodeContainer m_gateways; //!< Set of gateways to connect to this NS
+    std::list<std::pair<Ptr<NetDevice>, Ptr<Node>>>
```
To the best of my understanding we are storing P2P net devices here, why not using the P2PGwRegistration_t type here? Unless you plan to extend the support for generic NetDevices in the near future

nondetalle replied:
Yes, I'd like to be able to support any type of connection (Csma for instance) as in reality GW-NS traffic is transported by UDP/TCP or even higher layers (MQTT). In that way the NS can just use the IP of each GW and one NetDevice on its side, instead of having to track which NetDevice to use as it happens now with P2P.

pagmatt commented:
> Hello @pagmatt,
> 
> Here is a new version, I tried to make it as clean as I could get it on my own. I rebased it onto develop, but you can still easily check the latest changes because they are all in commit [9ab8421](https://github.com/signetlabdei/lorawan/commit/9ab84216db850700df7f0a133a06d3ac652ad0f3). I followed your suggestion to try and have a sort of `RegisterGatewayP2P` function to be used in the main simulation files. As I mentioned in my last reply, this moves a bit of complexity in the main files (see the examples) but I do not think it is too bad.
> 
> In summary, the changes proposed in this PR are meant to accomplish the following:
> 
> * Expose the P2P connection between GWs and server to prevent the bug that happened if the gateways' apps were installed before the server's app (highlighted in issue [m_receiveCallback in Receive function in lora-net-device #150](https://github.com/signetlabdei/lorawan/issues/150)). I feel it is also correct in theory because you should not be forced to have a specific order of installation of independent applications on different nodes;
> * Allow users to instantiate the P2P connection before installing applications, which is more in line with ns-3's way of doing things (i.e., that you should normally instantiate things in the order of nodes -> net devices -> applications, being completely separate steps);
> * Clarify that the module currently only supports P2P connections internally by forcing you to use `AddGatewayP2P` to register gateways connected to the network server.
> 
> Other minor changes coming with the last commit are:
> 
> * removing the option to provide a NodeContainer to NetworkServerHelper. Installing the app on multiple nodes does not make a lot of sense given that you have to provide net devices created beforehand.  In most cases one server is enough but people who want to play with more of them can always create different NetworkServerHelpers and register different devices/gateways to them;
> * adding a data structure to help storing what's needed for gateway registration in the NS app. It makes easier to have the P2P connection in one part of your code without depending on whether the NetworkServerHelper has been created yet.
> 
> Let me know your thoughts.

Apologies for the late reply. Changes are ok with me, great work! I just left a minor comment

nondetalle committed:
Update supported ns-3 version to ns-3.41 (#157)

* Patch code for ns-3.41 compatibility
* Update CI to replicate the one of ns-3.41

nondetalle commented:
Hello @pagmatt, I don't see your comment here

pagmatt commented:
> Hello @pagmatt, I don't see your comment here

My bad, the comment was still pending. You should be able to see it now

nondetalle commented on examples/adr-example.cc:
```diff
@@ -269,16 +269,31 @@ main(int argc, char* argv[])
     networkServerHelper.EnableAdr(adrEnabled);
     networkServerHelper.SetAdr(adrType);
-    networkServerHelper.Install(networkServers);
+    for (const auto& [serverP2PNetDev, gwNode] : gwRegistration)
+    {
+        networkServerHelper.AddGatewayP2P(serverP2PNetDev, gwNode);
+    }
```
I was thinking about directly passing a `P2PGwRegistration_t` collection to `AddGatewayP2P` and to move this loop inside the function. It cleans up the examples a little bit.

pagmatt replied:
I am in favor of this, even though I do not consider it blocking for this MR

nondetalle committed:
Directly pass P2PGwRegistration_t to NetworkServerHelper
